### PR TITLE
[#1786] Trackevents no longer overlap but inhabit the track below instea...

### PR DESCRIPTION
**Pull request by dseif**
...d or create a new one

### CODE REVIEW

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @ScottDowne received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/693dbf31c620176/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/693dbf31c620176/output.txt

Total script time: 0.75 mins

Server is running on http://pr683.butyr.org/templates/test.html

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @ScottDowne received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/4e1dc1bcefda0cf/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/4e1dc1bcefda0cf/output.txt

Total script time: 1.07 mins

Server is running on http://pr683.butyr.org/templates/test.html

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @ScottDowne received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/5dec80232031fe5/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/5dec80232031fe5/output.txt

Total script time: 1.29 mins

Server is running on http://pr683.butyr.org/templates/basic/
Run the Test Suite: http://pr683.butyr.org/test/

File: src/core/track.js [hunk 1]
```diff
@@ -50,7 +48,9 @@ define( [
 
     this.updateTrackEvents = function() {
       for ( var i = 0, l = _trackEvents.length; i < l; i++ ) {
-        _trackEvents[ i ].update();
+        if ( !_trackEvents[ i ].view.ghost ) {
```
Comment (ScottDowne) [inline src/core/track.js]: I wonder if it makes more sense to put this if inside the trackEvent.update. And if ghost, return early.

Comment (ScottDowne) [inline src/core/track.js]: Yeah, inside trackEvent.update, check _isGhost.

File: src/core/views/track-view.js [hunk 2]
```diff
@@ -24,43 +24,58 @@ define( [ "core/logger", "core/eventmanager", "util/dragndrop" ],
...
         drop: function( dropped, mousePosition ) {
+          var droppedElement = dropped.data ? dropped.data.element : dropped,
+              trackEventView,
+              track;
 
-          var draggableType = dropped.getAttribute( "data-butter-draggable-type" );
-
-          var start,
+          var draggableType = droppedElement.getAttribute( "data-butter-draggable-type" ),
+              start,
```
Comment (ScottDowne) [inline src/core/views/track-view.js:33]: Why is this in two seperate var blocks?

File: src/core/views/track-view.js [hunk 3]
```diff
@@ -127,21 +147,83 @@ define( [ "core/logger", "core/eventmanager", "util/dragndrop" ],
...
+        if ( trackEventView !== otherTrackEventView && !otherTrackEventView.dragging ) {
+          rect2 = otherTrackEventView.element.getBoundingClientRect();
+          // if a trackevent overlaps and it's not a ghost...
+          if ( !otherTrackEventView.isGhost && isOverlapping( rect1, rect2 ) ) {
+            return otherTrackEventView.trackEvent;
+          }
+        }
+      }
+      return null;
+    };
```
Comment (ScottDowne) [inline src/core/views/track-view.js:212]: These two functions, findOverlappingTrackEvent and findOverlappingTrackEventFromRect are far too similar.

Either remove findOverlappingTrackEvent outright. Or, update findOverlappingTrackEvent to simply call findOverlappingTrackEventFromRect, and pass in the rect.

I already see a bug that was fixed in findOverlappingTrackEvent that was not fixed in findOverlappingTrackEventFromRect.

in findOverlappingTrackEvent
if ( trackEventView !== otherTrackEventView && !otherTrackEventView.dragging ) {

and

findOverlappingTrackEvent
if ( !otherTrackEventView.dragging ) {

Should probably want to pull the findOverlappingTrackEvent version in to findOverlappingTrackEvent. Seems more complete.

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_check** from @humphd received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/1943d697ca7614c/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/1943d697ca7614c/output.txt

Total script time: 0.93 mins

File: src/core/eventmanager.js [hunk 4]
```diff
@@ -132,7 +141,7 @@ define( [], function(){
       e = {
         type: eventName.type,
         target: eventName.target,
-        data: eventName.data
+        data: eventName.data,
```
Comment (humphd) [inline src/core/eventmanager.js]: This comma is wrong.

File: src/core/track.js [hunk 5]
```diff
@@ -21,16 +19,16 @@ define( [
...
-        _view = new TrackView( this ),
+        _id = "" + __guid++,
+        _name = options.name || "Layer" + _id,
+        _view = new TrackView( _id, this ),
         _popcornWrapper = null,
         _this = this;
 
     _this._media = null;
     _this.order = 0;
+    _this.ghost = null;
```
Comment (humphd) [inline src/core/track.js:36]: Please add a comment to explain what this variable is for, since the name doesn't really indicate much.

File: src/core/trackevent.js [hunk 6]
```diff
@@ -41,7 +41,7 @@ define( [
    *
    * @param {Object} options: Options for initialization. Can contain the properties type, name, and popcornOptions. If the popcornOptions property is specified, its contents will be used to initialize the plugin instance associated with this TrackEvent.
```
Comment (humphd) [inline src/core/trackevent.js:42]: Please update this comment to reflect the fact that the options object is not the only thing passed in, and what isGhost means.

File: src/core/trackevent.js [hunk 7]
```diff
@@ -262,6 +270,21 @@ define( [
         }
       },
 
+      /*
+       * Property: isGhost
+       *
+       * Specifies whether this trackEvent is a ghost or not
```
Comment (humphd) [inline src/core/trackevent.js]: And being a ghost means what?  Needs more info somewhere on the whole "ghost" concept.

File: src/core/views/track-view.js [hunk 8]
```diff
@@ -24,43 +24,57 @@ define( [ "core/logger", "core/eventmanager", "util/dragndrop" ],
...
             });
           }
-          else if( draggableType === "trackevent" ) {
-            if( dropped.parentNode !== _element ){
-              left = dropped.offsetLeft;
+          else if ( draggableType === "trackevent" ) {
+            trackEventView = dropped.data;
+            track = trackEventView.trackEvent.track;
+
+            //Only rearrange trackEvent if it was moved *onto* this track
```
Comment (humphd) [inline src/core/views/track-view.js]: Space after //

File: src/core/views/track-view.js [hunk 9]
```diff
@@ -127,21 +146,66 @@ define( [ "core/logger", "core/eventmanager", "util/dragndrop" ],
...
       _this.unchain( trackEvent, [
         "trackeventmousedown",
         "trackeventmouseover",
         "trackeventmouseout"
       ]);
-    }; //removeTrackEvent
+    };
 
-  }; //TrackView
+    this.addTrackEventGhost = function( ghost ) {
```
Comment (humphd) [inline src/core/views/track-view.js:165]: Again, add some kind of overall discussion comment that explains what a ghost is.

File: src/core/views/trackevent-view.js [hunk 10]
```diff
@@ -56,6 +58,48 @@ define( [ "core/logger", "core/eventmanager", "util/dragndrop",
...
 
+    /**
+     * Member: createGhost
+     *
+     * Creates a clone of the current trackEvent that does not have an associated Popcorn trackevent.
+     * Used to notify the user when a trackevent overlaps and where the new location will be
+     * when the trackevent is dropped
+     */
+    this.createGhost = function( track ) {
+      if ( _ghost ) {
```
Comment (humphd) [inline src/core/views/trackevent-view.js:69]: Why throw?  Why not just return _ghost?

File: src/core/views/trackevent-view.js [hunk 11]
```diff
@@ -56,6 +58,48 @@ define( [ "core/logger", "core/eventmanager", "util/dragndrop",
...
+        element: clone
+      };
+
+      return _ghost;
+    };
+
+    /*
+     * Member: cleanupGhost
+     *
+     * Removes this trackEvents ghost and makes sure isGhost is set to false
```
Comment (humphd) [inline src/core/views/trackevent-view.js]: trackEvent's (apostrophe)

File: src/timeline/ghost-track.js [hunk 12]
```diff
@@ -0,0 +1,79 @@
+/* This Source Code Form is subject to the terms of the MIT license
+ * If a copy of the MIT license was not distributed with this file, you can
+ * obtain one at http://www.mozillapopcorn.org/butter-license.txt */
+
+define( [], function() {
```
Comment (humphd) [inline src/timeline/ghost-track.js:10]: Needs explanatory comment block at the front to indicate what this is all about.

File: src/timeline/trackevent-drag-manager.js [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+        correctOverlappingTrackEvents( trackEvent );
+      }
+
+      return !!newTrackEvent;
+    };
+
+    this.trackEventDropped = function( trackEvent, newTrack, startTime ) {
+      var newTrackEvent,
+          oldTrack = trackEvent.track,
+          corn = trackEvent.popcornOptions,
```
Comment (humphd) [inline src/timeline/trackevent-drag-manager.js]: change this to `popcornOptions` vs. corn.

File: src/util/dragndrop.js [hunk 14]
```diff
@@ -39,7 +39,7 @@ define([], function(){
     }
   }
 
-  function onDragged( e ){
+  var onDragged = function( e ){
```
Comment (humphd) [inline src/util/dragndrop.js]: Why this change?  Functions are better for memory.

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @humphd received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/6514bd2ce8152c0/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Failed

Full output at http://botio.butyr.org:8000/6514bd2ce8152c0/output.txt

Total script time: 2.00 mins

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_preview** from @humphd received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/8a39fbd4de3b827/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/8a39fbd4de3b827/output.txt

Total script time: 1.05 mins

Server is running on http://pr683.butyr.org/templates/basic/
Run the Test Suite: http://pr683.butyr.org/test/

**Comment by jbuck** [comment]
Staged on master: https://github.com/mozilla/butter/commit/987ff9b32e7b597acd671792d4641881791827b3
